# imports

In [57]:
import sys
sys.path.append('..')

In [58]:
from src.utils import load_data

In [59]:
fraud_df = load_data('../data/00_raw/Fraud_Data.csv')
credit_df = load_data('../data/00_raw/creditcard.csv')

In [60]:
fraud_df.head()

user_id          signup_time        purchase_time  ...  age    ip_address class
0    22058  2015-02-24 22:55:49  2015-04-18 02:47:11  ...   39  7.327584e+08     0
1   333320  2015-06-07 20:39:50  2015-06-08 01:38:54  ...   53  3.503114e+08     0
2     1359  2015-01-01 18:52:44  2015-01-01 18:52:45  ...   53  2.621474e+09     1
3   150084  2015-04-28 21:13:25  2015-05-04 13:54:50  ...   41  3.840542e+09     0
4   221365  2015-07-21 07:09:52  2015-09-09 18:40:53  ...   45  4.155831e+08     0

[5 rows x 11 columns]

In [61]:
fraud_df.isna().sum().sum()

np.int64(0)

In [62]:
fraud_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151112 entries, 0 to 151111
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   user_id         151112 non-null  int64  
 1   signup_time     151112 non-null  object 
 2   purchase_time   151112 non-null  object 
 3   purchase_value  151112 non-null  int64  
 4   device_id       151112 non-null  object 
 5   source          151112 non-null  object 
 6   browser         151112 non-null  object 
 7   sex             151112 non-null  object 
 8   age             151112 non-null  int64  
 9   ip_address      151112 non-null  float64
 10  class           151112 non-null  int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 12.7+ MB


In [63]:
fraud_df.describe()

user_id  purchase_value  ...    ip_address          class
count  151112.000000   151112.000000  ...  1.511120e+05  151112.000000
mean   200171.040970       36.935372  ...  2.152145e+09       0.093646
std    115369.285024       18.322762  ...  1.248497e+09       0.291336
min         2.000000        9.000000  ...  5.209350e+04       0.000000
25%    100642.500000       22.000000  ...  1.085934e+09       0.000000
50%    199958.000000       35.000000  ...  2.154770e+09       0.000000
75%    300054.000000       49.000000  ...  3.243258e+09       0.000000
max    400000.000000      154.000000  ...  4.294850e+09       1.000000

[8 rows x 5 columns]

In [64]:
fraud_df['class'].value_counts(normalize=True)

class
0    0.906354
1    0.093646
Name: proportion, dtype: float64

In [65]:
credit_df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [66]:
credit_df.isna().sum().sum()

np.int64(0)

In [67]:
credit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [68]:
credit_df.describe()

Time            V1  ...         Amount          Class
count  284807.000000  2.848070e+05  ...  284807.000000  284807.000000
mean    94813.859575  1.168375e-15  ...      88.349619       0.001727
std     47488.145955  1.958696e+00  ...     250.120109       0.041527
min         0.000000 -5.640751e+01  ...       0.000000       0.000000
25%     54201.500000 -9.203734e-01  ...       5.600000       0.000000
50%     84692.000000  1.810880e-02  ...      22.000000       0.000000
75%    139320.500000  1.315642e+00  ...      77.165000       0.000000
max    172792.000000  2.454930e+00  ...   25691.160000       1.000000

[8 rows x 31 columns]

## 📝 Data Understanding Summary

- 🧹 **Cleanliness:** Both datasets (`fraud_df` and `credit_df`) are very clean with no missing values.
- 🏷️ **Data Types:**  
    - In `fraud_df`, `signup_time` and `purchase_time` should be converted to datetime, and `ip_address` is stored as float but should be string/object.
    - In `credit_df`, all columns have appropriate numeric types.
- ⚖️ **Imbalance:**  
    - `fraud_df` has a fraud rate of about 1.5%, which is imbalanced but less extreme than some credit card datasets.
    - `credit_df` has a highly imbalanced class distribution, with fraud cases making up a very small fraction of the data (much less than 1%).
- 📊 **Size:**  
    - `fraud_df`: 151,112 rows × 11 columns.
    - `credit_df`: 284,807 rows × 31 columns.
- 🔍 **Features:**  
    - `fraud_df` contains user and transaction metadata.
    - `credit_df` contains anonymized principal components (V1–V28), `Time`, `Amount`, and `Class`.

## Part 1: Data Cleaning & Correction

### 1. Correcting Time-Based Data
**Hypothesis (If X...):** If we can measure the time between events, we can uncover behavioral patterns unique to fraudsters. The `signup_time` and `purchase_time` columns are currently stored as text (object), preventing any time-based calculations.

**Action (Then Y...):** Then we must convert these columns to a datetime format.

**Result (Therefore Z...):** Therefore, we unlock the ability to engineer crucial, time-sensitive features, such as the time between signup and purchase.

In [69]:
import pandas as pd
import numpy as np

# Convert to datetime
fraud_df['purchase_time'] = pd.to_datetime(fraud_df['purchase_time'])
fraud_df['signup_time'] = pd.to_datetime(fraud_df['signup_time'])

# Verify the change
fraud_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151112 entries, 0 to 151111
Data columns (total 11 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   user_id         151112 non-null  int64         
 1   signup_time     151112 non-null  datetime64[ns]
 2   purchase_time   151112 non-null  datetime64[ns]
 3   purchase_value  151112 non-null  int64         
 4   device_id       151112 non-null  object        
 5   source          151112 non-null  object        
 6   browser         151112 non-null  object        
 7   sex             151112 non-null  object        
 8   age             151112 non-null  int64         
 9   ip_address      151112 non-null  float64       
 10  class           151112 non-null  int64         
dtypes: datetime64[ns](2), float64(1), int64(4), object(4)
memory usage: 12.7+ MB


Now that the time-based columns are in the correct format, we can engineer a new feature: the time difference between signup and purchase. This can be a powerful indicator of fraudulent activity.

In [70]:
fraud_df['time_to_purchase'] = (fraud_df['purchase_time'] - fraud_df['signup_time']).dt.total_seconds()

# Display the new feature
fraud_df[['signup_time', 'purchase_time', 'time_to_purchase', 'class']].head()

signup_time       purchase_time  time_to_purchase  class
0 2015-02-24 22:55:49 2015-04-18 02:47:11         4506682.0      0
1 2015-06-07 20:39:50 2015-06-08 01:38:54           17944.0      0
2 2015-01-01 18:52:44 2015-01-01 18:52:45               1.0      1
3 2015-04-28 21:13:25 2015-05-04 13:54:50          492085.0      0
4 2015-07-21 07:09:52 2015-09-09 18:40:53         4361461.0      0

Let's visualize the distribution of `time_to_purchase` for both fraudulent and legitimate transactions to see if there are any discernible patterns.

In [71]:
import plotly.express as px

fig = px.histogram(
    fraud_df,
    x='time_to_purchase',
    color='class',
    nbins=100,
    log_y=True,
    title='Distribution of Time to Purchase by Class',
    labels={'time_to_purchase': 'Time from Signup to Purchase (seconds)'}
)
fig.show()

**Visual Analysis Summary:**

The histogram reveals a distinct pattern. A significant number of fraudulent transactions occur very quickly after signup, often within a day. This suggests that `time_to_purchase` is a valuable feature for our fraud detection model.

### 2. Standardizing IP Addresses
**Hypothesis (If X...):** If we can map a transaction's IP address to a country, we can identify high-risk geographic regions. The `ip_address` is currently a float, which could lead to precision errors and is incompatible for merging with the IP-to-country mapping table.

**Action (Then Y...):** Then we must convert the `ip_address` column to a 64-bit integer type, which is a standard, precise format for IP numbers.

**Result (Therefore Z...):** Therefore, the data is now prepared for a fast and accurate merge with the `IpAddress_to_Country.csv` dataset, enabling geolocation analysis.

In [72]:
# Convert ip_address to int64
fraud_df['ip_address'] = fraud_df['ip_address'].astype(np.int64)

# Verify the change
fraud_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151112 entries, 0 to 151111
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   user_id           151112 non-null  int64         
 1   signup_time       151112 non-null  datetime64[ns]
 2   purchase_time     151112 non-null  datetime64[ns]
 3   purchase_value    151112 non-null  int64         
 4   device_id         151112 non-null  object        
 5   source            151112 non-null  object        
 6   browser           151112 non-null  object        
 7   sex               151112 non-null  object        
 8   age               151112 non-null  int64         
 9   ip_address        151112 non-null  int64         
 10  class             151112 non-null  int64         
 11  time_to_purchase  151112 non-null  float64       
dtypes: datetime64[ns](2), float64(1), int64(5), object(4)
memory usage: 13.8+ MB


With the `ip_address` standardized, we can now enrich our data by mapping IP addresses to their respective countries. Let's load the IP-to-country mapping data and merge it with our fraud dataset.

In [73]:
# Load IP to Country data
ip_to_country_df = load_data('../data/00_raw/IpAddress_to_Country.csv')
ip_to_country_df.head()

lower_bound_ip_address  upper_bound_ip_address    country
0              16777216.0                16777471  Australia
1              16777472.0                16777727      China
2              16777728.0                16778239      China
3              16778240.0                16779263  Australia
4              16779264.0                16781311      China

In [74]:
# Merge dataframes to get country information
def map_ip_to_country(ip):
    try:
        return ip_to_country_df[(ip_to_country_df['lower_bound_ip_address'] <= ip) & (ip_to_country_df['upper_bound_ip_address'] >= ip)]['country'].iloc[0]
    except IndexError:
        return 'Unknown'

fraud_df['country'] = fraud_df['ip_address'].apply(map_ip_to_country)

fraud_df[['ip_address', 'country', 'class']].head()

ip_address        country  class
0   732758368          Japan      0
1   350311387  United States      0
2  2621473820  United States      1
3  3840542443        Unknown      0
4   415583117  United States      0

### A More Robust Approach: Bayesian Averaging for Fraud Rates

The challenge with picking a `min_transactions` threshold is that it's arbitrary. A better way to handle this is with **Bayesian Averaging**.

This method adjusts a country's calculated fraud rate by "shrinking" it towards the global average fraud rate. The amount of shrinkage depends on the number of transactions:
- **Countries with many transactions:** Their calculated rate is trusted, so it changes very little.
- **Countries with few transactions:** Their rate is unreliable, so it gets pulled more strongly towards the global average.

This gives us a more stable and reliable estimate of risk for each country.

In [75]:
# Calculate the global fraud rate
global_fraud_rate = fraud_df['class'].mean()

# Set a "credibility" parameter (C). This is like adding C "ghost" transactions with the global average rate.
# A common choice for C is the average number of transactions per country.
C = fraud_rate_df['total_transactions'].mean()

# Calculate the Bayesian-adjusted fraud rate
fraud_rate_df['bayesian_fraud_rate'] = (
    (fraud_rate_df['total_transactions'] / (fraud_rate_df['total_transactions'] + C)) * (fraud_rate_df['fraudulent_transactions'] / fraud_rate_df['total_transactions']) +
    (C / (fraud_rate_df['total_transactions'] + C)) * global_fraud_rate
) * 100 # Convert to percentage

# Sort by the new adjusted rate
top_bayesian_rate_countries = fraud_rate_df.sort_values(by='bayesian_fraud_rate', ascending=False).head(15)

print("Top 15 Countries by Bayesian-Adjusted Fraud Rate:")
print(top_bayesian_rate_countries[['total_transactions', 'fraudulent_transactions', 'fraud_rate', 'bayesian_fraud_rate']])

Top 15 Countries by Bayesian-Adjusted Fraud Rate:
              total_transactions  ...  bayesian_fraud_rate
country                           ...                     
New Zealand                  278  ...            12.609812
Ireland                      240  ...            12.403487
Denmark                      490  ...            11.796897
Luxembourg                    72  ...            11.720539
Saudi Arabia                 264  ...            11.674533
Tunisia                      118  ...            11.468352
Peru                         119  ...            11.456271
Chile                        417  ...            11.364898
Mexico                      1121  ...            11.364443
Ecuador                      106  ...            11.294923
Canada                      2975  ...            11.188457
Norway                       609  ...            10.891010
Sweden                      1090  ...            10.870921
Belgium                      409  ...            10.792729
Kuwait

In [76]:
# Visualize the top 15 countries by Bayesian-adjusted fraud rate
fig = px.bar(top_bayesian_rate_countries, 
             x=top_bayesian_rate_countries.index, 
             y='bayesian_fraud_rate',
             title='Top 15 Countries by Bayesian-Adjusted Fraud Rate',
             labels={'x': 'Country', 'y': 'Adjusted Fraud Rate (%)'},
             hover_data=['total_transactions', 'fraudulent_transactions', 'fraud_rate'])
fig.show()

**Bayesian Averaging Analysis Summary:**

This chart provides our most reliable view of country-level risk. Notice how some countries that had very high raw fraud rates (but low transaction counts) are no longer at the top. The countries that remain have a high fraud rate that is supported by a reasonable number of transactions.

This adjusted fraud rate will be a much more robust feature for our machine learning model.


## Part 2: Final Feature Engineering

### 1. Engineering Time-Based Patterns
**Hypothesis (If X...):** If fraudsters operate at specific times to avoid detection (e.g., late at night), then the hour of purchase will be a predictive signal.

**Action (Then Y...):** Then we will extract the hour of the day and the day of the week from the `purchase_time` column.

**Result (Therefore Z...):** Therefore, we create features that allow the model to learn time-based behavioral patterns.

In [77]:
# Extract hour of day and day of week
fraud_df['purchase_hour'] = fraud_df['purchase_time'].dt.hour
fraud_df['purchase_day_of_week'] = fraud_df['purchase_time'].dt.dayofweek # Monday=0, Sunday=6

fraud_df[['purchase_time', 'purchase_hour', 'purchase_day_of_week']].head()


purchase_time  purchase_hour  purchase_day_of_week
0 2015-04-18 02:47:11              2                     5
1 2015-06-08 01:38:54              1                     0
2 2015-01-01 18:52:45             18                     3
3 2015-05-04 13:54:50             13                     0
4 2015-09-09 18:40:53             18                     2

In [78]:
# Visualize fraud counts by hour of day
fig = px.histogram(fraud_df, 
                   x='purchase_hour', 
                   color='class',
                   barmode='group',
                   title='Fraud vs. Legit Transactions by Hour of Day',
                   labels={'purchase_hour': 'Hour of Day'})
fig.show()

### 2. Engineering Device-Based Anomaly Detection
**Hypothesis (If X...):** If a single physical device is used to access many different user accounts, it is highly indicative of a coordinated fraud ring.

**Action (Then Y...):** Then we will create a feature that counts how many unique users are associated with each `device_id`.

**Result (Therefore Z...):** Therefore, we create a feature that directly quantifies device sharing, a powerful signal for detecting organized fraud.

In [79]:
# Calculate number of unique users per device
device_user_counts = fraud_df.groupby('device_id')['user_id'].nunique().to_dict()
fraud_df['device_user_count'] = fraud_df['device_id'].map(device_user_counts)

fraud_df[['device_id', 'user_id', 'device_user_count', 'class']].sort_values('device_user_count', ascending=False).head(10)

device_id  user_id  device_user_count  class
123627  EQYVNEGOFLAWK    51141                 20      1
8829    EQYVNEGOFLAWK   386159                 20      1
80780   EQYVNEGOFLAWK     1985                 20      1
56906   EQYVNEGOFLAWK    57405                 20      1
77137   NGQCKIADMZORL   139414                 20      1
100762  KIPFSCNUGOLDP    45046                 20      1
35570   ITUMJCKWEYNDD    22163                 20      1
33712   ZUSVMDEZRBDTX   155117                 20      1
150265  CQTUVBYIWWWBC    46958                 20      1
116506  EQYVNEGOFLAWK   199435                 20      1

In [80]:
# Visualize the distribution of device_user_count for fraudulent transactions
fig = px.histogram(fraud_df[fraud_df['class']==1], 
                   x='device_user_count',
                   title='Distribution of Device Sharing for Fraudulent Transactions',
                   labels={'device_user_count': 'Number of Users per Device'})
fig.show()

## Part 3: Finalizing the Dataset for Modeling
Our feature engineering is now complete. The final step before modeling is to clean up the dataframe, keeping only the features we will use.

In [81]:
# Add the Bayesian-adjusted fraud rate to the dataframe
fraud_rate_df.rename(columns={'bayesian_fraud_rate': 'country_risk_rate'}, inplace=True)
fraud_df = fraud_df.merge(fraud_rate_df['country_risk_rate'], on='country', how='left')

# Drop identifier/redundant columns
final_df = fraud_df.drop(columns=[
    'user_id', 
    'signup_time', 
    'purchase_time', 
    'device_id', 
    'ip_address',
    'country' # Drop original country column after merging risk rate
])

# One-hot encode categorical variables
final_df = pd.get_dummies(final_df, columns=['source', 'browser', 'sex'], drop_first=False)

print("Final dataset columns:")
print(final_df.columns)
final_df.head()

Final dataset columns:
Index(['purchase_value', 'age', 'class', 'time_to_purchase', 'purchase_hour',
       'purchase_day_of_week', 'device_user_count', 'country_risk_rate',
       'country_risk_rate', 'source_Ads', 'source_Direct', 'source_SEO',
       'browser_Chrome', 'browser_FireFox', 'browser_IE', 'browser_Opera',
       'browser_Safari', 'sex_F', 'sex_M'],
      dtype='object')


purchase_value  age  class  ...  browser_Safari  sex_F  sex_M
0              34   39      0  ...           False  False   True
1              16   53      0  ...           False   True  False
2              15   53      1  ...           False  False   True
3              44   41      0  ...            True  False   True
4              39   45      0  ...            True  False   True

[5 rows x 19 columns]

In [85]:
final_df.drop_duplicates(inplace=True)

In [90]:
final_df = final_df.loc[:, ~final_df.columns.duplicated(keep='first')]

Now we save the final, fully processed dataset.

In [93]:
from src.utils import save_data
# Save the final processed dataframe
save_data(final_df, '../data/01_processed/ecommerce_fraud_processed.csv')

Data saved to ../data/01_processed/ecommerce_fraud_processed.csv


### scalling data for creditcard

In [95]:
from sklearn.preprocessing import StandardScaler

# Create a scaler object
scaler = StandardScaler()

In [97]:
# Fit and transform the 'Amount' and 'Time' columns
credit_df['scaled_amount'] = scaler.fit_transform(credit_df['Amount'].values.reshape(-1, 1))
credit_df['scaled_time'] = scaler.fit_transform(credit_df['Time'].values.reshape(-1, 1))

In [98]:
# Drop the original, unscaled columns
credit_df.drop(['Time', 'Amount'], axis=1, inplace=True)

In [100]:
credit_df.head()

V1        V2        V3  ...  Class  scaled_amount  scaled_time
0 -1.359807 -0.072781  2.536347  ...      0       0.244964    -1.996583
1  1.191857  0.266151  0.166480  ...      0      -0.342475    -1.996583
2 -1.358354 -1.340163  1.773209  ...      0       1.160686    -1.996562
3 -0.966272 -0.185226  1.792993  ...      0       0.140534    -1.996562
4 -1.158233  0.877737  1.548718  ...      0      -0.073403    -1.996541

[5 rows x 31 columns]

In [99]:
save_data(credit_df, '../data/01_processed/creditcard_processed.csv')

Data saved to ../data/01_processed/creditcard_processed.csv
